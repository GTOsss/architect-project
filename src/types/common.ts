export type AnyObject = Record<string, unknown>;
export type AnyMethod = (...rest: any[]) => any;
