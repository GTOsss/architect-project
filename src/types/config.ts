export type TemplateConfig = {
  replace: boolean;
  clean: boolean;
  watch: boolean;
  output: string;
  itrStart: string;
  itrEnd: string;
  itrFileNameStart: string;
  itrFileNameEnd: string;
  templateExt: string;
  esLint: {
    quiet: boolean;
  };
};

export type ArcConfig = TemplateConfig & {
  templates: Record<string, Partial<TemplateConfig>>;
};
