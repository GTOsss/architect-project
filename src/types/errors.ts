export enum ErrorNameEnum {
  sourceMapValueHasNotTemplateName = 'sourceMapValueHasNotTemplateName',
}
