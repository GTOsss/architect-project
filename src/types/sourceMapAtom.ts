/** Template params in "source-map-atom" */
export type TemplateParamsMA = {
  /** Component name (not template) */
  name?: string;
  /** Relative project path */
  path?: string;
  /** Relative default path (from default params) */
  rPath?: string;
};

export type DefaultParamsMA = Record<string, TemplateParamsMA>;

export type TemplateSourceMapMAValue = string | [string, TemplateParamsMA];

export type SourceMapAtom = Record<string, Array<string | [string, TemplateParamsMA]>>;

export type SourceMapAtomRequiredFile = {
  aliases: Record<string, string>;
  defaultParams: DefaultParamsMA;
  map: SourceMapAtom;
};
