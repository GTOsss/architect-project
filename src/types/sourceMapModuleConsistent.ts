export type TemplateParamsConsistent = {
  /** Template name */
  template: string;
  /** Value of variable name for template.
   * @example
   * // In this example name = "button":
   * { [button]: { template: 'react-component' } } */
  name: string;
  /** Output path describe in source map as object key in first lvl source map.
   * @example
   * // In this example outputPath = 'some/output/path'
   * const map = {
   *    'some/output/path': { [button]: template: 'react-component' }
   * }
   * */
  outputPath: string;
  /** Some custom template params from source-map */
  [key: string]: any;
};

/** Second lvl (file-name to template lvl) source-map-module in **consistent format**.
 * @Example
 *
 * const sourceMap = {
 *   <...>,
 *
 *   'src/components/inputs': {
 *     Button: {
 *       template: 'react-component',
 *     }
 *     index: {
 *       template: 'index-file',
 *       params: { <params for "index-file" template> },
 *     },
 *   },
 *
 *   <...>,
 * }
 * */
export type FileLevelConsistentTemplateMap = {
  [name: string]: TemplateParamsConsistent;
};

/**
 * Top level (dir-path lvl) source-map-module in **consistent format**.
 * Pair { "path": { <FileLevelConsistentTemplateMap> } }
 *
 * @Example
 * const sourceMap = {
 *   'src/components': { <FileLevelConsistentTemplateMap> },
 *   'src/user/components': { <FileLevelConsistentTemplateMap> },
 *   'src/pages': { <FileLevelConsistentTemplateMap> },
 *   ...
 * };
 *
 * */
export type SourceMapModuleConsistent = {
  /**
   * "key" = path-to-dir
   *
   * "value" = { <FileLevelTemplateMap> } */
  [path: string]: FileLevelConsistentTemplateMap;
};

export type SourceMapModuleConsistentRequiredFile = {
  aliases: Record<string, string>;
  map: SourceMapModuleConsistent;
};
