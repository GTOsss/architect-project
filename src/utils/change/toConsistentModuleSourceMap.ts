import {
  isString,
  isTemplateAsArray,
  isTemplateAsObj,
  TemplateParamsMMAsObj,
  TemplateSourceMapValue,
} from '../../types/sourceMapModule';

/**
 * @example
 * 'react-component'              →   {template: 'react-component'}
 * ['react-component', {}]        →   {template: 'react-component'}
 * {template: 'react-component'}  →   {template: 'react-component'}
 * */
const transformToConsistentTemplateValue = (templateValue: TemplateSourceMapValue, aliases: Record<string, string>) => {
  let templateName: string;
  let templateParams: TemplateParamsMMAsObj = {};

  if (isString(templateValue)) {
    templateName = templateValue;
  } else if (isTemplateAsArray(templateValue)) {
    templateName = templateValue[0];
    templateParams = templateValue[1];
  } else if (isTemplateAsObj(templateValue)) {
    templateName = templateValue.template;
    delete templateValue.template;
    templateParams = templateValue;
  }

  templateName = aliases[templateName] || templateName;

  return { template: templateName, ...templateParams };
};

/**
 * @example
 * const sourceMap = {
 *   button1: 'react-component'
 *   button2: ['react-component', {}]
 *   button3: {template: 'react-component'}
 * };
 *
 * ↓  ↓  ↓
 *
 * {
 *   button1: {template: 'react-component'},
 *   button2: {template: 'react-component'},
 *   button3: {template: 'react-component'}
 * };
 * */
export const toConsistentModuleSourceMap = ({ map: sourceMapModule2 = {}, aliases = {} }) => {
  const moduleMap = Object.entries(sourceMapModule2).reduce((acc, [path, fileToTemplateMap]) => {
    acc[path] = Object.entries(fileToTemplateMap).reduce((acc, [fileName, templateValue]) => {
      acc[fileName] = transformToConsistentTemplateValue(templateValue, aliases);
      return acc;
    }, {});

    return acc;
  }, {});

  return { map: moduleMap, aliases };
};
