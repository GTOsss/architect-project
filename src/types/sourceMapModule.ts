import { ErrorNameEnum } from './errors';

/** Template params in "source-map-module" as arr */
export type TemplateParamsMMAsArray = [string, { [key: string]: any }];

/** Template params in "source-map-module" as obj */
export type TemplateParamsMMAsObj = {
  /** Template name */
  template?: string;
  [key: string]: any;
};

export type TemplateSourceMapValue = string | TemplateParamsMMAsObj | TemplateParamsMMAsArray;

// todo need test and review for this type
// export type TemplateAsObj = {
//   [key: string]: string[];
// };

/**
 * Second level (file-name to template-name lvl) source-map-module.
 * Pair { ["fileName"]: <templateName/templateParamsObj> }
 * @Example:
 *
 * const sourceMap = {
 *   <...>,
 *
 *   'src/components/inputs': {
 *     Button: 'react-component'
 *     Input: 'react-component',
 *     index: {
 *       template: 'index-file',
 *       params: { <params for "index-file" template> },
 *     },
 *   },
 *
 *   <...>,
 * }
 *
 * */
export type FileLevelTemplateMap = {
  [name: string]: TemplateSourceMapValue;
};

/**
 *  Top level (dir-path lvl) source-map-module.
 *  Pair { "path": { FileLevelTemplateMap> } }
 *
 *  @Example:
 *
 * const sourceMap = {
 *   'src/components': { <FileLevelTemplateMap> },
 *   'src/user/components': { <FileLevelTemplateMap> },
 *   'src/pages': { <FileLevelTemplateMap> },
 *   ...
 * };
 *
 * */
export type SourceMapModule = {
  /**
   * "key" = path-to-dir
   *
   * "value" = { <FileLevelTemplateMap> } */
  [path: string]: FileLevelTemplateMap;
};

export const isString = (value: any): value is string => typeof value === 'string';

export const isTemplateAsArray = (value: TemplateSourceMapValue): value is TemplateParamsMMAsArray => {
  if (isString(value)) {
    return false;
  }

  return Array.isArray(value) && typeof value[0] === 'string';
};

export const isTemplateAsObj = (value: TemplateSourceMapValue): value is TemplateParamsMMAsObj => {
  if (isString(value)) {
    return false;
  }

  return !Array.isArray(value) && typeof value.template === 'string';
};

/**
 * Will throw error if sourceMap has incorrect value.
 * */
export const checkValidTemplateValue = (value: TemplateSourceMapValue) => {
  if (isString(value) || isTemplateAsObj(value) || isTemplateAsArray(value)) {
    return;
  }

  const err = new Error();
  err.name = ErrorNameEnum.sourceMapValueHasNotTemplateName;
  throw err;
};
