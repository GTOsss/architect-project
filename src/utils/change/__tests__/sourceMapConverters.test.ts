import { SourceMapModule } from '../../../types/sourceMapModule';
import { toConsistentModuleSourceMap } from '../toConsistentModuleSourceMap';

const aliases = {
  /** react-component template */
  rc: 'react-component',

  /** index template - will create "barrel file" via _script_ */
  i: 'index',
};

describe('Source map converters', () => {
  it('module2ToModule simple (only string values)', () => {
    const sourceMapSimple: SourceMapModule = {
      'src/components': {
        header: 'rc',
      },
      'src/components/inputs': {
        button: 'rc',
        index: 'i',
      },
    };

    const result = toConsistentModuleSourceMap({ map: sourceMapSimple, aliases });
    expect(result).toMatchSnapshot();
  });

  it('module2ToModule simple (string and object values)', () => {
    const sourceMapSimple: SourceMapModule = {
      'src/components': {
        header: 'rc',
      },
      'src/components/inputs': {
        button: 'rc',
        index: { template: 'i' },
      },
    };

    const result = toConsistentModuleSourceMap({ map: sourceMapSimple, aliases });
    expect(result).toMatchSnapshot();
  });

  it('module2ToModule simple (string and array values)', () => {
    const sourceMapSimple: SourceMapModule = {
      'src/components': {
        header: 'rc',
      },
      'src/components/inputs': {
        button: 'rc',
        index: ['i', {}],
      },
    };

    const result = toConsistentModuleSourceMap({ map: sourceMapSimple, aliases });
    expect(result).toMatchSnapshot();
  });
});
