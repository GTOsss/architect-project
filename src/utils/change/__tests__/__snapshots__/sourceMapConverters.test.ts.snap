// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Source map converters module2ToModule simple (only string values) 1`] = `
{
  "aliases": {
    "i": "index",
    "rc": "react-component",
  },
  "map": {
    "src/components": {
      "header": {
        "template": "react-component",
      },
    },
    "src/components/inputs": {
      "button": {
        "template": "react-component",
      },
      "index": {
        "template": "index",
      },
    },
  },
}
`;

exports[`Source map converters module2ToModule simple (string and array values) 1`] = `
{
  "aliases": {
    "i": "index",
    "rc": "react-component",
  },
  "map": {
    "src/components": {
      "header": {
        "template": "react-component",
      },
    },
    "src/components/inputs": {
      "button": {
        "template": "react-component",
      },
      "index": {
        "template": "index",
      },
    },
  },
}
`;

exports[`Source map converters module2ToModule simple (string and object values) 1`] = `
{
  "aliases": {
    "i": "index",
    "rc": "react-component",
  },
  "map": {
    "src/components": {
      "header": {
        "template": "react-component",
      },
    },
    "src/components/inputs": {
      "button": {
        "template": "react-component",
      },
      "index": {
        "template": "index",
      },
    },
  },
}
`;
